File: des--elev_rud_model_TT_z_v_psi_cpp_ws/src/auv_imp/src/auv_imp.cpp
#include <ros/ros.h>
#include <avl_bsd_translator/CommandMsg.h>
#include <avl_bsd_translator/ActionMsg.h>
#include <avl_bsd_translator/HeartbeatMsg.h>
#include <std_msgs/Float64.h>
#include <cmath>

class AUVNode {
public:
    AUVNode(int argc, char** argv) : rate(1.0) {
        ros::init(argc, argv, "auv_imp");

        heartbeat_pub = nh.advertise<avl_bsd_translator::HeartbeatMsg>("bsd/heartbeat", 10);
        z_pub = nh.advertise<std_msgs::Float64>("bsd/z", 10);
        command_sub = nh.subscribe("bsd/command", 10, &AUVNode::command_callback, this);
        action_sub = nh.subscribe("bsd/action", 10, &AUVNode::action_callback, this);

        elevator = 0.0;
        rudder = 0.0;
        rpm = 0.0;
        z = 250.0;  // Initial depth value
        x = 51.5074;    // Initial x position (latitude)
        y = -0.1278;    // Initial y position (longitude)
    }

    void command_callback(const avl_bsd_translator::CommandMsg::ConstPtr& command_msg) {
        ROS_INFO("command_msg received");
    }

    void action_callback(const avl_bsd_translator::ActionMsg::ConstPtr& action_msg) {
        elevator = std::stod(action_msg->VALUES[1]);
        rudder = std::stod(action_msg->VALUES[2]);
        rpm = std::stod(action_msg->VALUES[3]);
    }

    void update_heartbeat() {
        double roll = 0.0;
        double pitch = elevator; // Use the elevator angle directly as the pitch value
        double yaw = rudder;     // Use the rudder angle directly as the yaw value
        double z_value = z;      // Set the initial depth value
        double u = 0.0;
        double v = 0.0;
        double w = 0.0;
        double p = 0.0;
        double q = 0.0;
        double r = 0.0;

        double dt = 1.0; // Time step in seconds

        // Convert RPM to revolutions per second (RPS)
        double rpm_to_rps = rpm / 60.0;

        // Assuming a propeller circumference of 0.1 meters
        double prop_circumference = 0.1;

        // Calculate approximate linear speed based on propeller circumference
        double linear_speed = rpm_to_rps * prop_circumference;

        // Calculate the change in x and y positions based on the approximate linear speed, pitch, and yaw
        double x_change = linear_speed * dt * cos(pitch) * cos(yaw);
        double y_change = linear_speed * dt * cos(pitch) * sin(yaw);

        // Convert the change in meters to latitude and longitude units (approximation for small area)
        double lat_to_meter = 1.0 / 111320.0;
        double lon_to_meter = 1.0 / (111320.0 * cos(x * M_PI / 180.0));

        // Update x and y positions in latitude and longitude units
        x += x_change * lat_to_meter;
        y += y_change * lon_to_meter;

        // Calculate the depth based on elevator control
        double elevator_rad = M_PI * elevator / 180.0;
        double depth = z_value + (w * dt) + (0.5 * elevator_rad * pow(dt, 2));

        z_value = depth; // Update the depth with the new calculated value

        z = z_value; // Store the updated depth value

        // Publish depth value to additional topic
        std_msgs::Float64 depth_msg;
        depth_msg.data = z;
        z_pub.publish(depth_msg);

        avl_bsd_translator::HeartbeatMsg heartbeat_msg;
        heartbeat_msg.HEARTBEAT_STATUS = 1;
        heartbeat_msg.HEARTBEAT_ROLL = roll;
        heartbeat_msg.HEARTBEAT_PITCH = pitch;
        heartbeat_msg.HEARTBEAT_YAW = yaw;
        heartbeat_msg.HEARTBEAT_VN = u;
        heartbeat_msg.HEARTBEAT_VE = v;
        heartbeat_msg.HEARTBEAT_VD = w;
        heartbeat_msg.HEARTBEAT_DEPTH = z;
        heartbeat_msg.HEARTBEAT_HEIGHT = y;
        heartbeat_msg.HEARTBEAT_RPM = rpm;
        heartbeat_msg.HEARTBEAT_WX = p;
        heartbeat_msg.HEARTBEAT_WY = q;
        heartbeat_msg.HEARTBEAT_WZ = r;
        heartbeat_msg.HEARTBEAT_AX = 0.0;
        heartbeat_msg.HEARTBEAT_AY = 0.0;
        heartbeat_msg.HEARTBEAT_AZ = 0.0;
        heartbeat_msg.HEARTBEAT_VOLTAGE = 0.0;
        heartbeat_msg.HEARTBEAT_UMODEM_SYNC = 0;
        heartbeat_msg.HEARTBEAT_MISSION_MODE = 0;
        heartbeat_msg.HEARTBEAT_FSD_MISSION_STATE = 0;
        heartbeat_msg.HEARTBEAT_FSD_CURRENT_ACTION = 0;
        heartbeat_msg.HEARTBEAT_FSD_TOTAL_ACTIONS = 0;
        heartbeat_msg.HEARTBEAT_FSD_ACTION_PERCENT = 0.0;
        heartbeat_msg.HEARTBEAT_BSD_MISSION_STATE = 0;
        heartbeat_msg.HEARTBEAT_BSD_CURRENT_ACTION = 0;
        heartbeat_msg.HEARTBEAT_BSD_TOTAL_ACTIONS = 0;
        heartbeat_msg.HEARTBEAT_BSD_ACTION_PERCENT = 0;
        heartbeat_msg.HEARTBEAT_GPS_SATS = 0;
        heartbeat_msg.HEARTBEAT_GPS_LAT = x;
        heartbeat_msg.HEARTBEAT_GPS_LON = y;
        heartbeat_msg.HEARTBEAT_GPS_ALT = 0.0;
        heartbeat_msg.HEARTBEAT_NAV_LAT = 0;
        heartbeat_msg.HEARTBEAT_NAV_LON = 0;
        heartbeat_msg.HEARTBEAT_NAV_ALT = 0.0;
        heartbeat_msg.HEARTBEAT_NAV_YAW_STD = 0;
        heartbeat_msg.HEARTBEAT_NAV_AVG_POS_ERR = 0;
        heartbeat_msg.HEARTBEAT_NAV_INITIALIZED = 0;
        heartbeat_pub.publish(heartbeat_msg);
    }

    void run() {
        while (ros::ok()) {
            update_heartbeat();
            rate.sleep();
            ros::spinOnce();
        }
    }

private:
    ros::NodeHandle nh;
    ros::Publisher heartbeat_pub;
    ros::Publisher z_pub;
    ros::Subscriber command_sub;
    ros::Subscriber action_sub;
    double elevator;
    double rudder;
    double rpm;
    double z;
    double x;
    double y;
    ros::Rate rate; // Add the 'rate' member variable here
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "auv_imp_cpp");
    AUVNode auv_node(argc, argv);
    auv_node.run();
    return 0;
}

File: des--elev_rud_model_TT_z_v_psi_py_ws/src/auv_imp/nodes/auv_imp.py
#!/usr/bin/env python3

import rospy
from avl_bsd_translator.msg import CommandMsg, ActionMsg, HeartbeatMsg
from std_msgs.msg import Float64
import math

class AUVNode:
    def __init__(self):
        rospy.init_node('auv_imp')

        self.heartbeat_pub = rospy.Publisher('bsd/heartbeat', HeartbeatMsg, queue_size=10)
        self.z_pub = rospy.Publisher('bsd/z', Float64, queue_size=10)  # New topic for z
        self.command_sub = rospy.Subscriber('bsd/command', CommandMsg, self.command_callback)
        self.action_sub = rospy.Subscriber('bsd/action', ActionMsg, self.action_callback)

        self.elevator = 0.0
        self.rudder = 0.0
        self.rpm = 0.0
        self.z = 250.0  # Initial depth value
        self.x = 0.0  # Initial x position
        self.y = 0.0  # Initial y position

        self.rate = rospy.Rate(1)  # Set the publishing frequency to 1 Hz

        self.update_heartbeat()  # Start publishing heartbeat messages

    def command_callback(self, command_msg):
        print("command_msg received")

    def action_callback(self, action_msg):
        self.elevator = float(action_msg.VALUES[1])
        self.rudder = float(action_msg.VALUES[2])
        self.rpm = float(action_msg.VALUES[3])

    def update_heartbeat(self):
        roll = 0.0
        pitch = 0.0
        yaw = 0.0
        z = self.z  # Set the initial depth value
        u = self.elevator
        v = self.rudder
        w = 0.0
        p = 0.0
        q = 0.0
        r = 0.0

        dt = 1.0  # Time step in seconds

        # Convert elevator and rudder from degrees to radians
        elevator_rad = math.radians(self.elevator)
        rudder_rad = math.radians(self.rudder)

        # Calculate depth based on elevator control
        depth = z + (w * dt) + (0.5 * elevator_rad * (dt ** 2))

        # Calculate the change in x and y positions based on rudder input
        x_change = v * dt * math.cos(rudder_rad)  # v is the rudder input
        y_change = v * dt * math.sin(rudder_rad)  # v is the rudder input

        # Update x and y positions
        self.x += x_change
        self.y += y_change

        z = depth  # Update the depth with the new calculated value

        self.z = z  # Store the updated depth value

        # Publish depth value to additional topic
        self.z_pub.publish(self.z)

        heartbeat_msg = HeartbeatMsg()
        heartbeat_msg.HEARTBEAT_STATUS = 1
        heartbeat_msg.HEARTBEAT_ROLL = roll
        heartbeat_msg.HEARTBEAT_PITCH = pitch
        heartbeat_msg.HEARTBEAT_YAW = yaw
        heartbeat_msg.HEARTBEAT_VN = u
        heartbeat_msg.HEARTBEAT_VE = v
        heartbeat_msg.HEARTBEAT_VD = w
        heartbeat_msg.HEARTBEAT_DEPTH = z
        heartbeat_msg.HEARTBEAT_HEIGHT = self.y
        heartbeat_msg.HEARTBEAT_RPM = self.rpm
        heartbeat_msg.HEARTBEAT_WX = p
        heartbeat_msg.HEARTBEAT_WY = q
        heartbeat_msg.HEARTBEAT_WZ = r
        heartbeat_msg.HEARTBEAT_AX = 0.0
        heartbeat_msg.HEARTBEAT_AY = 0.0
        heartbeat_msg.HEARTBEAT_AZ = 0.0
        heartbeat_msg.HEARTBEAT_VOLTAGE = 0.0
        heartbeat_msg.HEARTBEAT_UMODEM_SYNC = 0
        heartbeat_msg.HEARTBEAT_MISSION_MODE = 0
        heartbeat_msg.HEARTBEAT_FSD_MISSION_STATE = 0
        heartbeat_msg.HEARTBEAT_FSD_CURRENT_ACTION = 0
        heartbeat_msg.HEARTBEAT_FSD_TOTAL_ACTIONS = 0
        heartbeat_msg.HEARTBEAT_FSD_ACTION_PERCENT = 0.0
        heartbeat_msg.HEARTBEAT_BSD_MISSION_STATE = 0
        heartbeat_msg.HEARTBEAT_BSD_CURRENT_ACTION = 0
        heartbeat_msg.HEARTBEAT_BSD_TOTAL_ACTIONS = 0
        heartbeat_msg.HEARTBEAT_BSD_ACTION_PERCENT = 0
        heartbeat_msg.HEARTBEAT_GPS_SATS = 0
        heartbeat_msg.HEARTBEAT_GPS_LAT = 0
        heartbeat_msg.HEARTBEAT_GPS_LON = 0
        heartbeat_msg.HEARTBEAT_GPS_ALT = 0.0
        heartbeat_msg.HEARTBEAT_NAV_LAT = 0
        heartbeat_msg.HEARTBEAT_NAV_LON = 0
        heartbeat_msg.HEARTBEAT_NAV_ALT = 0.0
        heartbeat_msg.HEARTBEAT_NAV_YAW_STD = 0
        heartbeat_msg.HEARTBEAT_NAV_AVG_POS_ERR = 0
        heartbeat_msg.HEARTBEAT_NAV_INITIALIZED = 0

        self.heartbeat_pub.publish(heartbeat_msg)

        self.rate.sleep()
        self.update_heartbeat()

if __name__ == '__main__':
    auv_node = AUVNode()
    rospy.spin()
